spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/StructInternalRow.java

```diff
@@ -356,4 +357,23 @@ private <T> GenericArrayData fillArray(
 
     return new GenericArrayData(array);
   }
+
+  @Override
+  public boolean equals(Object other) {
```

Comment:
Had to implement as Spark collects grouping keys and checks their equality.

Reply:
What is the contract for equals for `InternalRow`? This probably works, but does it indicate a bug in Spark where Spark assumes that `InternalRow` implementations consider one another equal?

Reply:
It is a bug in Spark that will be fixed. Spark uses `groupBy` in one place, which relies on equals.
All other places use `InternalRowSet`.